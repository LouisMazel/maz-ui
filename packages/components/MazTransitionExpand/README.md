# MazTransitionExpand

## Slots

<!-- @vuese:MazTransitionExpand:slots:start -->

| Name    | Description | Default Slot Content |
| ------- | ----------- | -------------------- |
| default | -           | -                    |

<!-- @vuese:MazTransitionExpand:slots:end -->
