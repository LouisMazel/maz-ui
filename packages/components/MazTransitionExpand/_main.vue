<template>
  <transition
    class="maz-transition-expand"
    name="maz-expand"
    @enter="enter"
    @after-enter="afterEnter"
    @leave="leave"
  >
    <slot />
  </transition>
</template>

<script>
export default {
  name: 'MazTransitionExpand',
  methods: {
    enter (element) {
      const width = getComputedStyle(element).width

      element.style.width = width
      element.style.position = 'absolute'
      element.style.visibility = 'hidden'
      element.style.height = 'auto'
      element.style.top = 'bottom'

      const height = getComputedStyle(element).height

      element.style.width = null
      element.style.position = null
      element.style.visibility = null
      element.style.height = 0

      // Force repaint to make sure the
      // animation is triggered correctly.
      /* eslint-disable */
        getComputedStyle(element).height
        /* eslint-enable */

      // Trigger the animation.
      // We use `setTimeout` because we need
      // to make sure the browser has finished
      // painting after setting the `height`
      // to `0` in the line above.
      this.$nextTick(() => {
        element.style.height = height
      })
    },
    afterEnter (element) {
      element.style.height = 'auto'
    },
    leave (element) {
      const height = getComputedStyle(element).height
      element.style.height = height

      // Force repaint to make sure the
      // animation is triggered correctly.
      /* eslint-disable */
        getComputedStyle(element).height
        /* eslint-enable */
      this.$nextTick(() => {
        element.style.height = 0
      })
    }
  }
}
</script>
