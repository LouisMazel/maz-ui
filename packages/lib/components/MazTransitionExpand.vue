<template>
  <TransitionGroup
    class="m-transition-expand"
    name="maz-expand"
    tag="div"
    @enter="(el: Element) => enter(el as HTMLElement)"
    @after-enter="(el: Element) => afterEnter(el as HTMLElement)"
    @leave="(el: Element) => leave(el as HTMLElement)"
  >
    <slot></slot>
  </TransitionGroup>
</template>

<script lang="ts" setup>
  import { nextTick } from 'vue'

  withDefaults(
    defineProps<{
      // Expand animation duration in milliseconds
      animationDuration?: string
    }>(),
    {
      animationDuration: '300ms',
    },
  )

  const enter = (element: HTMLElement) => {
    const width = getComputedStyle(element).width

    element.style.width = width
    element.style.position = 'absolute'
    element.style.visibility = 'hidden'
    element.style.height = 'auto'
    element.style.top = 'bottom'

    const height = getComputedStyle(element).height

    element.style.width = ''
    element.style.position = ''
    element.style.visibility = ''
    element.style.height = '0px'

    // Force repaint to make sure the
    // animation is triggered correctly.
    getComputedStyle(element).height

    // Trigger the animation.
    // We use `setTimeout` because we need
    // to make sure the browser has finished
    // painting after setting the `height`
    // to `0` in the line above.
    nextTick(() => {
      element.style.height = height
    })
  }

  const afterEnter = (element: HTMLElement) => {
    element.style.height = 'auto'
  }

  const leave = (element: HTMLElement) => {
    const height = getComputedStyle(element).height
    element.style.height = height

    // Force repaint to make sure the
    // animation is triggered correctly.
    getComputedStyle(element).height

    nextTick(() => {
      element.style.height = '0px'
    })
  }
</script>

<style lang="postcss" scoped>
  .m-transition-expand,
  .m-transition-expand * {
    will-change: height;
    transform: translateZ(0);
    backface-visibility: hidden;
    perspective: 1000px;

    :deep(> *) {
      @apply maz-duration-[v-bind(animationDuration)];
    }
  }
</style>
