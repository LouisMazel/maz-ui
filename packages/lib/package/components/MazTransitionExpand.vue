<template>
  <Transition
    class="m-transition-expand"
    name="maz-expand"
    tag="div"
    @enter="(el) => enter(el as HTMLElement)"
    @after-enter="(el) => afterEnter(el as HTMLElement)"
    @leave="(el) => leave(el as HTMLElement)"
  >
    <slot />
  </Transition>
</template>

<script lang="ts" setup>
  import { nextTick, ref } from 'vue'

  enum States {
    NONE = 'none',
    ENTERING = 'entering',
    AFTER_ENTERING = 'entering',
    LEAVING = 'entering',
  }
  const state = ref<States>(States.NONE)

  const enter = (element: HTMLElement) => {
    if (state.value !== States.NONE) {
      return
    }

    const width = getComputedStyle(element).width

    element.style.width = width
    element.style.position = 'absolute'
    element.style.visibility = 'hidden'
    element.style.height = 'auto'
    element.style.top = 'bottom'

    const height = getComputedStyle(element).height

    element.style.width = ''
    element.style.position = ''
    element.style.visibility = ''
    element.style.height = '0px'

    // Force repaint to make sure the
    // animation is triggered correctly.
    /* eslint-disable */
    getComputedStyle(element).height
    /* eslint-enable */

    // Trigger the animation.
    // We use `setTimeout` because we need
    // to make sure the browser has finished
    // painting after setting the `height`
    // to `0` in the line above.
    nextTick(() => {
      element.style.height = height
    })
  }
  const afterEnter = (element: HTMLElement) => {
    if (state.value !== States.NONE) {
      return
    }

    element.style.height = 'auto'
  }
  const leave = (element: HTMLElement) => {
    if (state.value !== States.NONE) {
      return
    }

    const height = getComputedStyle(element).height
    element.style.height = height

    // Force repaint to make sure the
    // animation is triggered correctly.
    /* eslint-disable */
    getComputedStyle(element).height
    /* eslint-enable */
    nextTick(() => {
      element.style.height = '0px'
    })
  }
</script>

<style lang="postcss" scoped>
  .m-transition-expand,
  .m-transition-expand * {
    will-change: height;
    transform: translateZ(0);
    backface-visibility: hidden;
    perspective: 1000px;
  }
</style>
