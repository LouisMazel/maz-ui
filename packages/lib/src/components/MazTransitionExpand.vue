<script lang="ts" setup>
import { nextTick } from 'vue'

export interface MazTransitionExpandProps {
  /** Duration of the animation in milliseconds */
  animationDuration?: string
}

withDefaults(defineProps<MazTransitionExpandProps>(), {
  animationDuration: '300ms',
})

function enter(element: HTMLElement) {
  const width = getComputedStyle(element).width

  element.style.width = width
  element.style.position = 'absolute'
  element.style.visibility = 'hidden'
  element.style.height = 'auto'
  element.style.top = 'bottom'

  const height = getComputedStyle(element).height

  element.style.width = ''
  element.style.position = ''
  element.style.visibility = ''
  element.style.height = '0px'

  // Force repaint to make sure the
  // animation is triggered correctly.

  // eslint-disable-next-line ts/no-unused-expressions
  getComputedStyle(element).height

  // Trigger the animation.
  // We use `setTimeout` because we need
  // to make sure the browser has finished
  // painting after setting the `height`
  // to `0` in the line above.
  nextTick(() => {
    element.style.height = height
  })
}

function afterEnter(element: HTMLElement) {
  element.style.height = 'auto'
}

function leave(element: HTMLElement) {
  const height = getComputedStyle(element).height
  element.style.height = height

  // Force repaint to make sure the
  // animation is triggered correctly.

  // eslint-disable-next-line ts/no-unused-expressions
  getComputedStyle(element).height

  nextTick(() => {
    element.style.height = '0px'
  })
}
</script>

<template>
  <TransitionGroup
    class="m-transition-expand m-reset-css"
    name="maz-expand"
    tag="div"
    @enter="(el: Element) => enter(el as HTMLElement)"
    @after-enter="(el: Element) => afterEnter(el as HTMLElement)"
    @leave="(el: Element) => leave(el as HTMLElement)"
  >
    <slot />
  </TransitionGroup>
</template>

<style lang="postcss" scoped>
  .m-transition-expand,
.m-transition-expand * {
  will-change: height;
  transform: translateZ(0);
  backface-visibility: hidden;
  perspective: 1000px;

  :deep(> *) {
    transition-duration: v-bind('animationDuration');
  }
}
</style>
