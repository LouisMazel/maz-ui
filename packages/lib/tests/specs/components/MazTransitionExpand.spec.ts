import MazTransitionExpand from '@components/MazTransitionExpand.vue'
import { shallowMount } from '@vue/test-utils'

describe('components/MazTransitionExpand.vue', () => {
  expect(MazTransitionExpand).toBeTruthy()

  const wrapper = shallowMount(MazTransitionExpand)

  it('should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
