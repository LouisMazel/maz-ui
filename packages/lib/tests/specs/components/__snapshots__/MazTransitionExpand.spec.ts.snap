// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/MazTransitionExpand.vue > should match with the snapshot 1`] = `"<transition-group-stub data-v-76cbd327="" name="maz-expand" tag="div" appear="false" persisted="false" css="true" class="m-transition-expand m-reset-css"></transition-group-stub>"`;
